keps/sig-scheduling/4832-async-preemption/README.md

```diff
@@ -0,0 +1,417 @@
+
+But, we have the similar problem with the preemption; the preemption is run at PostFilter extension point which is the part of the scheduling cycle.
+The preemption has to make some API calls to update Pods' condition and delete Pods after all, which could block the scheduling throughput.
+
+scheduler-perf [actually shows](https://github.com/kubernetes/kubernetes/blob/342da505bdefbd849b808cca3cb76c24a993025f/test/integration/scheduler_perf/config/performance-config.yaml#L641) currently the preemption scenario takes too long time, compared to others.
+
+### Goals
+
+- Improve scheduling throughput when pods require issuing preemptions by making API calls asynchronous
+
+### Non-Goals
+
+- Making the same enhancement for DRA is not a goal of this KEP because it's an under-construction feature yet.
+  - If DRA maintainers want, technically they can along with this KEP. But, at least in this KEP, we don't discuss how.
+
+## Proposal
+
+The preemption plugin makes API calls for the preemption asynchronously after `PostFilter` extension point 
+so that the scheduler can continue to other Pods' scheduling while making API calls for preemption.
+After the preemption goroutine is done, the scheduling for the Pod that triggered the preemption will be retried.
+
+### Risks and Mitigations
+
+#### When kube-apiserver is unstable
+
+When kube-apiserver is unstable and API calls at the preemption goroutine fails frequently,
+the scheduler could make a non-optimal scheduling decision 
+because the scheduler nominates pods at `PostFilter` though, those Pods won't be scheduled on nodes because the preemption API calls fail.
+
+Let's say many mid-priority Pods are making the preemption API calls.
+With the scheduler after this proposal, during the preemption goroutine for them are runnning, 
+the scheduler assumes they'll be scheduled at the Nodes eventually
+that the preemptions are targeting via `.Status.NominatedNodeName`.
+So, other mid-priority or lower priority Pods' scheduling take those preemptor Pods into consideration, 
+which is correct if the preemption goroutine finishes successful actually, while which results in non-best scheduling results otherwise.
+(Higher priority Pods won't be affected; Pods can take place of reserved for lower priority Pods via `.Status.NominatedNodeName`)
+
+But, in the first place though, when kube-apiserver is unstable, the scheduler doesn't behave well 
+because it works with a lot of communication with kube-apiserver.
+Even if the scheduler makes the best scheduling result, the binding API might fail after all.
+
+So, we don't have to pay a special attention to this issue.
+
+## Design Details
+
+To achieve an asynchronous preemption, we will change the preemption plugin's implementation like the following:
+1. The preemption PostFilter plugin calculates the preemption target and nominate the Pod for the Node. (We'll use `AddNominatedPod` API exposed from the scheduling framework to plugins.)
```

Comment:
What will happen once we restart the scheduler, we'll loss the whole nominator.

Reply:
For the Pods that have completed the preemption API calls completely, they have `.status.nominatedNodeName` 
and hence wouldn't be affected by the restart.

But, regarding the Pods that are in-progress of making API calls, the nomination info for them would be lost,
and the deletion of victim Pods are, of course, stopped in the middle.
After the restart, such Pods would be rescheduled;
the deletion of victims by the preemptions before the restart will likely be helpful in making them schedulable. 
And also, some of them will likely have to trigger the preemption again though, 
those Pods are likely to pick up the same node as the preemption target 
especially because those nodes lost some running Pods already because of the previous preemption(s). 
(unless there's a lot of change in the cluster during the restart, which should be a short duration).

Yes, a lot of assumption ("likely") in this explanation though, given the restart is not often, we believe we can ignore it.

Reply:
Thanks for the explanation.